misc: fix memory leaks

rr- opened the pull request:
The worst offender was the audio samples loader that only freed first 20 samples instead of all of them, which leaked tons of raw PCM data for each level start / savegame load. The second worst offender was the gameflow loader which didn't do anything at all to free the memory.

Also added `Memory_Dup` that replaces `strdup` so that we only allocate with `Memory_Alloc` across the entire application. This has actually helped me track bad allocations (see the notes below).
Also added `Memory_FreePointer` utility that can replaces the following idiom:
```c
if (memory) {
    Memory_Free(memory);
    memory = NULL;
}
```
with
```c
Memory_FreePointer(&memory);
```

---

The following patch and script helped me analyze the leaks. It tracks every single memory operation in logs, and the script checks if the memory is still reachable. Adding these in case this ever needs to be used again. In theory the same should be doable with valgrind, but I can assure you that using valgrind on wine is not a pleasant experience.

```diff
diff --git a/src/memory.c b/src/memory.c
index c73c24cf..582fe29c 100644
--- a/src/memory.c
+++ b/src/memory.c
@@ -1,14 +1,20 @@
 #include "memory.h"
 
 #include "game/shell.h"
+#include "log.h"
 
 #include <assert.h>
 #include <stdlib.h>
 #include <string.h>
 
-void *Memory_Alloc(size_t size)
+#define DEBUG_MEMORY
+
+void *Memory_AllocImpl(size_t size, const char *file, int line, const char *func)
 {
     void *result = malloc(size);
+    #ifdef DEBUG_MEMORY
+    Log_Message(file, line, func, "Allocating memory (%d bytes): %p", size, result);
+    #endif
     if (!result) {
         Shell_ExitSystem("ERROR: Could not allocate enough memory");
     }
@@ -16,33 +22,39 @@ void *Memory_Alloc(size_t size)
     return result;
 }
 
-void *Memory_Realloc(void *memory, size_t size)
+void *Memory_ReallocImpl(void *memory, size_t size, const char *file, int line, const char *func)
 {
     void *result = realloc(memory, size);
+    #ifdef DEBUG_MEMORY
+    Log_Message(file, line, func, "Reallocating memory at %p (%d bytes): %p", memory, size, result);
+    #endif
     if (!result) {
         Shell_ExitSystem("ERROR: Could not allocate enough memory");
     }
     return result;
 }
 
-void Memory_Free(void *memory)
+void Memory_FreeImpl(void *memory, const char *file, int line, const char *func)
 {
+    #ifdef DEBUG_MEMORY
+    Log_Message(file, line, func, "Freeing memory at %p", memory);
+    #endif
     if (memory) {
         free(memory);
     }
 }
 
-void Memory_FreePointer(void *arg)
+void Memory_FreePointerImpl(void *arg, const char *file, int line, const char *func)
 {
     void *memory;
-    memcpy(&memory, arg, sizeof(void *));
-    memcpy(arg, &(void *) { NULL }, sizeof(void *));
-    Memory_Free(memory);
+    memcpy(&memory, arg, sizeof(void*));
+    memcpy(arg, &(void *) { NULL }, sizeof(void*));
+    Memory_FreeImpl(memory, file, line, func);
 }
 
-void *Memory_Dup(const char *string)
+void *Memory_DupImpl(const char *string, const char *file, int line, const char *func)
 {
-    void *memory = Memory_Alloc(strlen(string) + 1);
+    void *memory = Memory_AllocImpl(strlen(string) + 1, file, line, func);
     strcpy(memory, string);
     return memory;
 }
diff --git a/src/memory.h b/src/memory.h
index 23c23f95..768b2a40 100644
--- a/src/memory.h
+++ b/src/memory.h
@@ -5,8 +5,14 @@
 
 #include <stdint.h>
 
-void *Memory_Alloc(size_t size);
-void *Memory_Realloc(void *memory, size_t size);
-void Memory_Free(void *memory);
-void Memory_FreePointer(void *memory);
-void *Memory_Dup(const char *string);
+#define Memory_Alloc(size) (Memory_AllocImpl(size, __FILE__, __LINE__, __func__))
+#define Memory_Realloc(memory, size) (Memory_ReallocImpl(memory, size, __FILE__, __LINE__, __func__))
+#define Memory_Free(memory) (Memory_FreeImpl(memory, __FILE__, __LINE__, __func__))
+#define Memory_FreePointer(memory) (Memory_FreePointerImpl(memory, __FILE__, __LINE__, __func__))
+#define Memory_Dup(string) (Memory_DupImpl(string, __FILE__, __LINE__, __func__))
+
+void *Memory_AllocImpl(size_t size, const char *file, int line, const char *func);
+void *Memory_ReallocImpl(void *memory, size_t size, const char *file, int line, const char *func);
+void Memory_FreeImpl(void *memory, const char *file, int line, const char *func);
+void Memory_FreePointerImpl(void *memory, const char *file, int line, const char *func);
+void *Memory_DupImpl(const char *string, const char *file, int line, const char *func);
```

```python3
#!/usr/bin/env python3
import re
from collections.abc import Iterable
from dataclasses import dataclass
from enum import Enum
from pathlib import Path


class OperationType(Enum):
    Allocation = 1
    Reallocation = 2
    Deallocation = 3


@dataclass
class Operation:
    file: str
    line: int
    func: str
    type: OperationType
    address: int
    new_address: int | None = None


@dataclass
class Problem:
    message: str
    operations: list[Operation]


def get_operations_from_log_lines(lines: Iterable[str]) -> Iterable[Operation]:
    for line in lines:
        line = line.decode('utf-8')
        if match := re.search(
            r"(\S+) (\d+) (\S+) Allocating memory .*: ([A-F0-9]*)", line
        ):
            yield Operation(
                file=match.group(1),
                line=int(match.group(2)),
                func=match.group(3),
                type=OperationType.Allocation,
                address=int(match.group(4), 16),
            )
        elif match := re.search(
            r"(\S+) (\d+) (\S+) Reallocating memory at ([A-F0-9]*) .*: ([A-F0-9]*)",
            line,
        ):
            yield Operation(
                file=match.group(1),
                line=int(match.group(2)),
                func=match.group(3),
                type=OperationType.Reallocation,
                address=int(match.group(4), 16),
                new_address=int(match.group(5), 16),
            )
        elif match := re.search(
            r"(\S+) (\d+) (\S+) Freeing memory at ([A-F0-9]*)", line
        ):
            yield Operation(
                file=match.group(1),
                line=int(match.group(2)),
                func=match.group(3),
                type=OperationType.Deallocation,
                address=int(match.group(4), 16),
            )


def find_problems(operations: Iterable[Operation]) -> Iterable[Problem]:
    reachable_memory: set[int] = set()
    for operation in operations:
        if operation.type == OperationType.Allocation:
            reachable_memory.add(operation.address)
        elif operation.type == OperationType.Reallocation:
            if operation.address != 0:
                reachable_memory.remove(operation.address)
            reachable_memory.add(operation.new_address)
        elif operation.type == OperationType.Deallocation:
            if operation.address != 0:
                try:
                    reachable_memory.remove(operation.address)
                except KeyError:
                    yield Problem("Invalid free", operations=[operation])

    for address in reachable_memory:
        yield Problem(
            f"Unfreed memory: {address:08x}",
            operations=[
                operation
                for operation in operations
                if operation.address == address
                or operation.new_address == address
            ],
        )


with Path("Tomb1Main.log").open('rb') as handle:
    operations = list(get_operations_from_log_lines(handle))
    problems = list(find_problems(operations))
    for problem in problems:
        print(problem.message, problem.operations[0].file, problem.operations[0].line)
```

walkawayy commented:
Very cool.

Looks like there are a few `strdup` calls left:
`json.c` line 2796
`context.c` line 241
`s_fmv.c` line 2097

There's also a bunch of `Memory_Free`'s left. For example in `level.c`:
```
    Memory_Free(sample_offsets);
    Memory_Free(sample_pointers);
    Memory_Free(sample_sizes);
```

Did you overlook those or is there a reason they didn't get updated to the new convention?

rr- commented:
Replaced the rest of `Memory_Free` with `Memory_FreePointer`. You're right, we're more consistent this way.

Good catch with context.c, fixed.

In s_fmv.c we use `av_strdup` which is part of libavcodec's ffplay codebase that we forked.
In json.c we use `malloc` which is part of some public domain JSON library that we forked.

I kept these two instances to be close to the original, but I have no strong opinion on this. If you tell me to replace these allocators with our routines I'll do so :)

walkawayy commented:
LGTM.

walkawayy commented:
Hm last comment. Maybe change the two `Memory_Free`s in `s_picture.c` to just make it fully consistent.
